File: 23-Graph/python/07_detect_a_cycle_in_directed_graph_using_bfs.py
from collections import deque
from typing import List

class Solution:
    def isCyclic(self, N: int, adj: List[List[int]]) -> bool:
        indegree = [0] * N
        for i in range(N):
            for neighbor in adj[i]:
                indegree[neighbor] += 1

        queue = deque([i for i in range(N) if indegree[i] == 0])
        count = 0

        while queue:
            node = queue.popleft()
            count += 1
            for neighbor in adj[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)

        return count != N

File: 23-Graph/cpp/07-detect-a-cycle-in-directed-graph-using-bfs.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    bool isCyclic(int N, vector<vector<int>>& adj) {
        vector<int> indegree(N, 0);
        for (int i = 0; i < N; i++) {
            for (int neighbor : adj[i]) {
                indegree[neighbor]++;
            }
        }

        queue<int> q;
        for (int i = 0; i < N; i++) {
            if (indegree[i] == 0) {
                q.push(i);
            }
        }

        int count = 0;
        while (!q.empty()) {
            int node = q.front();
            q.pop();
            count++;
            for (int neighbor : adj[node]) {
                indegree[neighbor]--;
                if (indegree[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }

        return count != N;
    }
};
